### ISSUE TITLE
What happen when all losses turn to nan????

### ISSUE THREAD

**Issue by chen-erqi** [KindXiaoming/pykan#396]
I use the latest pykan. Multikan is amazing to fit more complicted physical problems.
But when I use the multiKan, I find a weired problem.
The multiKan seems to turn to nan easily.
Given a dataset and a multiKan, it works well.
But when I reduce the sample number of this dataset to feed into a same multiKan, all losses turn to nan.
Could anyone please to figure out this problem?
thanks a lot!!!

**Comment by fmunzlin** [comment]
First of all, I do not run multiKan but rather their basic version, but I run into this issue on a daily.  I guess the underlying issue is that we haven't found a reference architecture for KAN's on a research specific topic and those are just the syndromes of it being very new. Remember we did a lot of research using non-linear activation functions, comparitively very little on learning them.  
To me, this sounds like as if the model is just sensitive to changes in at least this hyperparameter. Changes in hyperparameters often have weird consequences. For me, NaN in losses often come from some layer where the multiplication of weights with some input exceed datatype. this leaves you with inf and running an operation on that turns NaN. At some point these will propagates through the network and into the loss. 
Try different learning rate settings / optimizers (AdamW >> LBFGS [working on genomics]), play with network size (smaller >> bigger, in this regard) and hope for the best.

**Comment by chen-erqi** [comment]
> First of all, I do not run multiKan but rather their basic version, but I run into this issue on a daily. I guess the underlying issue is that we haven't found a reference architecture for KAN's on a research specific topic and those are just the syndromes of it being very new. Remember we did a lot of research using non-linear activation functions, comparitively very little on learning them. To me, this sounds like as if the model is just sensitive to changes in at least this hyperparameter. Changes in hyperparameters often have weird consequences. For me, NaN in losses often come from some layer where the multiplication of weights with some input exceed datatype. this leaves you with inf and running an operation on that turns NaN. At some point these will propagates through the network and into the loss. Try different learning rate settings / optimizers (AdamW >> LBFGS [working on genomics]), play with network size (smaller >> bigger, in this regard) and hope for the best.

Thanks for your reply!
Your reply is very useful for my question and it is same as my experience.

But the "nan" is bad for the explainability of Kan.
We need to figure out the difference between a Kan with or without some samples.

In my case, changing update grid to false could solve this question, I hope it can help you.
In my opinion, now, the nan in Kan is mostly related to the coding rather than the general rules in mlps or cnns.

**Comment by fmunzlin** [comment]
don't get me wrong. I'm not saying you should "accept" that. I'll well aware that having nan's in your weight will at some point result in problems at least doing inference. What I am saying is try different settings for both you experiments AND its counterpart until you find a setting that works for both of them. 
honestly, I don't use updating the grid at all anymore. tried to make it work but it seems to buggy atm. same for expanding the grid

**Comment by chen-erqi** [comment]
> don't get me wrong. I'm not saying you should "accept" that. I'm well aware that having nan's in your weight will at some point result in problems at least doing inference. What I am saying is try different settings for both you experiment AND its counterpart until you find a setting that works for both of them. honestly, I don't use updating the grid at all anymore. tried to make it work but it seems to buggy atm. same for expanding the grid

hhh I know your meaning. Please forgive my poor English.
Generally, changing setting for better performance is ok, but in some case such as comparing the difference between two datasets, setting need to be constant.
And I sincerely recommend the update grid, the improvement by that is remarkable.
